lib/bricks-net-http-server.js

```diff
@@ -0,0 +1,295 @@
+
+var makeCppException = require('./cpp-exceptions').makeCppException;
+var HTTPException = require('./bricks-net-exceptions').HTTPException;
+var CannotServeStaticFilesOfUnknownMIMEType = require('./bricks-net-exceptions').CannotServeStaticFilesOfUnknownMIMEType;
+var AttemptedToSendHTTPResponseMoreThanOnce = require('./bricks-net-exceptions').AttemptedToSendHTTPResponseMoreThanOnce;
+var HandlerAlreadyExistsException = makeCppException(HTTPException, "HandlerAlreadyExistsException");
+var HandlerDoesNotExistException = makeCppException(HTTPException, "HandlerDoesNotExistException");
+
+var DefaultFourOhFourMessage = require('./bricks-net-http-default_messages').DefaultFourOhFourMessage;
+var DefaultInternalServerErrorMessage = require('./bricks-net-http-default_messages').DefaultInternalServerErrorMessage;
+var DefaultMethodNotAllowedMessage = require('./bricks-net-http-default_messages').DefaultMethodNotAllowedMessage;
+
+var HTTPResponseCode = require('./bricks-net-http-codes').HTTPResponseCode;
+
+var Socket = require('./bricks-net-tcp-impl').Socket;
+
+var HTTPServerConnection = require('./bricks-net-http-impl').HTTPServerConnection;
+var ConnectionClose = require('./bricks-net-http-impl').ConnectionClose;
+var ConnectionKeepAlive = require('./bricks-net-http-impl').ConnectionKeepAlive;
+
+var HTTPHeaders = require('./bricks-net-http-headers').HTTPHeaders;
+
+
+function StaticFileServer(body, content_type) {
+	return function (r) {
+		if (r.method == "GET") {
+			r.connection.SendHTTPResponse(body, HTTPResponseCode.OK, content_type);
+		}
+		else {
+			r.connection.SendHTTPResponse(
+				DefaultMethodNotAllowedMessage(), HTTPResponseCode.MethodNotAllowed, "text/html");
+		}
+	};
+}
+
+
+function Request(connection) {
+	if (!(this instanceof Request)) {
+		return new Request(connection);
+	}
+	
+	// Return a function that mimics the C++ `operator()` overload.
+	var _this = function () {
+		_this._DEBUG_LOG(_this + '#operator()(' + [].map.call(arguments, function (arg) {
+			return require('util').inspect(arg);
+		}).join(', ') + ')');
+		return _this.connection.SendHTTPResponse.apply(connection, arguments);
+	};
+	
+	Object.defineProperties(_this, {
+		_DEBUG_LOG: {
+			configurable: true,
+			enumerable: false,
+			value:
+				/* istanbul ignore next: debug method */
+				function () {}
+				//function () { console.log.apply(console, arguments); }
+		},
+		toString: {
+			configurable: true,
+			enumerable: false,
+			value: function () {
+				return 'Request(' + connection + ')';
+			}
+		}
+	});
+	
+	_this.SendChunkedResponse = function () {
+		return _this.connection.SendChunkedHTTPResponse();
+	};
+	
+	_this.connection = connection;
+	var http_data = _this.http_data = connection.HTTPRequest();
+	_this.url = http_data.URL();
+	_this.method = http_data.Method();
+	var has_body = _this.has_body = http_data.HasBody();
+	var empty_string = _this.empty_string = "";
+	_this.body = (has_body ? http_data.Body() : empty_string);
```

Comment:
> This is totally unnecessary in JS. The empty_string helper is only there in C++ because body should be a const reference.

As long as Bricks code defines this field as public, I have to mimic it.